### DISCUSSION TITLE
Handling Negation in the Dependency Matcher

### DISCUSSION THREAD

**Discussion by Fourthought** [explosion/spaCy#7446] [Help: Coding & Implementations] [answered]
Hi There,

Does anyone have any ideas for how to handle negation with the dependency matcher.

For example, the following phrase can be identified with the following pattern:

`

    "my friends are here"

    pattern = [
        {"RIGHT_ID": "PREDICATE", "RIGHT_ATTRS": {"DEP": "AUX"}}, 
        {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "SUBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}}, 
        {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}}
    ]
`

but creating a negation pattern generates hits for both the original and negated pattern
`

    "my friends are not here"
        negated_pattern = [
        {"RIGHT_ID": "PREDICATE", "RIGHT_ATTRS": {"DEP": "AUX"}}, 
        {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "SUBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}}, 
        {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}}
        {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"DEP": "neg"}}
    ]
`

Is there any way to isolate the original and negated patterns?

**Comment by polm** [comment]
Your patterns as written didn't work for me, but with a little reworking it seems to work the way you want it to. This pattern matches "My friends are not here" but not "My friends are here":

```
pattern = [
    {"RIGHT_ID": "PREDICATE", "RIGHT_ATTRS": {"POS": "AUX"}}, 
    {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "SUBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}}, 
    {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}},
    {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "NEGATION", "RIGHT_ATTRS": {"DEP": "neg"}}
]
```

Full code:

```
import spacy
from spacy.matcher import DependencyMatcher

nlp = spacy.load("en_core_web_sm")
matcher = DependencyMatcher(nlp.vocab)

pattern = [
    {"RIGHT_ID": "PREDICATE", "RIGHT_ATTRS": {"POS": "AUX"}}, 
    {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "SUBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}}, 
    {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}},
    {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "NEGATION", "RIGHT_ATTRS": {"DEP": "neg"}}
]

matcher.add("NOTHERE", [negated_pattern])

texts = [
        "My friends are not here",
        "My friends are here",
        ]

for text in texts:
    print()
    doc = nlp(text)
    for word in doc:
        print(word, word.pos_, word.dep_)
    matches = matcher(doc)
    print(matches)
    if matches:
        match_id, token_ids = matches[0]
        for i in range(len(token_ids)):
            print(pattern[i]["RIGHT_ID"] + ":", doc[token_ids[i]].text)
```

**Comment by Fourthought** [comment, marked as answer]
Thank you for the response!

Sorry, I'd not explained myself well, I've modified the code to explain it better.

There are two patterns: the first is a straightforward pattern, while the second is a negated pattern. Below, you'll see that the negation sentence creates two matches one for the straightforward pattern and a second for the negation pattern.

I would like to only have a negation pattern match for the negated sentence. Is this possible?

`

    import spacy
    from spacy.matcher import DependencyMatcher

    nlp = spacy.load("en_core_web_sm")
    matcher = DependencyMatcher(nlp.vocab)

    patterns = [
        {"label": "HERE",
         "pattern": [
             {"RIGHT_ID": "PREDICATE", "RIGHT_ATTRS": {"POS": "AUX"}},
             {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "SUBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}},
             {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"POS": "ADV"}},
         ]},
         {"label": "NOTHERE",
         "pattern": [
             {"RIGHT_ID": "PREDICATE", "RIGHT_ATTRS": {"POS": "AUX"}},
             {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "SUBJECT", "RIGHT_ATTRS": {"POS": "NOUN"}},
             {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "OBJECT", "RIGHT_ATTRS": {"POS": "ADV"}},
             {"LEFT_ID": "PREDICATE", "REL_OP": ">", "RIGHT_ID": "NEGATION", "RIGHT_ATTRS": {"DEP": "neg"}}
         ]}]

    for pattern in patterns:
        matcher.add(pattern["label"], [pattern["pattern"]])

    texts = [
            "My friends are not here",
            "My friends are here",
            ]

    for text in texts:
        print()
        print(text)
       doc = nlp(text)
       for word in doc:
        print(word, word.pos_, word.dep_)
    matches = matcher(doc)
    print(matches)
    if matches:
        print()

        for match in matches:
            match_id, token_ids = match # get the match_id and match tokens
            match_id = nlp.vocab.strings[match_id] # get the match_id name
            pattern = matcher.get(match_id)[1][0] # get the match pattern from the matcher
            
            for i in range(len(token_ids)):
                print(pattern[i]["RIGHT_ID"] + ":", doc[token_ids[i]].text)
                print()

`

**Reply by polm** [reply]
Ah, OK, I see what you mean. I don't think that's possible with the Dependency Matcher as it stands now - there's no negation operator, so you can't do something like match only non-negated structures. 

What I would do to capture this is to use both rules, and if both match in a sentence you can compare the matches and see if the negative one is just the positive one plus a negation, and then remove the positive one. It's a little redundant but not that complicated.

**Reply by Fourthought** [reply]
Thank you Paul, 

I've managed a workaround based on what you've suggested, so thank you.

Having an option for a dependency could be quite useful and cut down on the number of rules required to parse language.

Steve

**Reply by polm** [reply]
The Dependency Matcher is pretty new, so I'd expect it to add features over time, and this seems like a compelling one - thanks for outlining a concrete use case for us to reference!
